journalbeat/input/input.go

```diff
@@ -0,0 +1,174 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package input
+
+import (
+	"fmt"
+	"sync"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/elastic/beats/journalbeat/checkpoint"
+	"github.com/elastic/beats/journalbeat/reader"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Input manages readers and forwards entries from journals.
+type Input struct {
+	readers  []*reader.Reader
+	done     chan struct{}
+	config   Config
+	pipeline beat.Pipeline
+	states   map[string]checkpoint.JournalState
+	id       uuid.UUID
+	logger   *logp.Logger
+}
+
+// New returns a new Inout
+func New(
+	c *common.Config,
+	pipeline beat.Pipeline,
+	done chan struct{},
+	states map[string]checkpoint.JournalState,
+) (*Input, error) {
+	config := DefaultConfig
+	if err := c.Unpack(&config); err != nil {
+		return nil, err
+	}
+
+	id := uuid.NewV4()
+	logger := logp.NewLogger("input").With("id", id)
+
+	var readers []*reader.Reader
+	if len(config.Paths) == 0 {
+		cfg := reader.Config{
+			Path:          reader.LocalSystemJournalID, // used to identify the state in the registry
+			Backoff:       config.Backoff,
+			MaxBackoff:    config.MaxBackoff,
+			BackoffFactor: config.BackoffFactor,
+			Seek:          config.Seek,
+		}
+
+		state := states[reader.LocalSystemJournalID]
+		r, err := reader.NewLocal(cfg, done, state, logger)
+		if err != nil {
+			return nil, fmt.Errorf("error creating reader for local journal: %v", err)
+		}
+		readers = append(readers, r)
+	}
+
+	for _, p := range config.Paths {
+		cfg := reader.Config{
+			Path:          p,
+			Backoff:       config.Backoff,
+			MaxBackoff:    config.MaxBackoff,
+			BackoffFactor: config.BackoffFactor,
+			Seek:          config.Seek,
+		}
+		state := states[p]
+		r, err := reader.New(cfg, done, state, logger)
+		if err != nil {
+			return nil, fmt.Errorf("error creating reader for journal: %v", err)
+		}
+		readers = append(readers, r)
+	}
+
+	logger.Debugf("New input is created for paths %v", config.Paths)
+
+	return &Input{
+		readers:  readers,
+		done:     done,
+		config:   config,
+		pipeline: pipeline,
+		states:   states,
+		id:       id,
+		logger:   logger,
+	}, nil
+}
+
+// Run connects to the output, collects entries from the readers
+// and then publishes the events.
+func (i *Input) Run() {
+	client, err := i.pipeline.ConnectWith(beat.ClientConfig{
+		PublishMode:   beat.GuaranteedSend,
+		EventMetadata: common.EventMetadata{},
+		Meta:          nil,
+		Processor:     nil,
+		ACKCount: func(n int) {
+			i.logger.Infof("journalbeat successfully published %d events", n)
+		},
+	})
+	if err != nil {
+		i.logger.Error("Error connecting to output: %v", err)
+		return
+	}
+	defer client.Close()
+
+	for {
+		select {
+		case <-i.done:
+			return
+		default:
+			i.publishAll(client)
+		}
+	}
+
+}
```

Comment:
Instead of doing a for loop here and check for `i.done` and calling `i.publishAll` we should just call `i.publishAll`. This would simplify the code logic.